### PR TITLE
Implement VectorTools::interpolate_to_finer/coarser_mesh().

### PR DESCRIPTION
**Pull request by bangerth**
I mentioned this on the elements forum: I'm in need of a function that can interpolate between coarse and fine `DoFHandler` objects not only on a single process, but indeed in parallel where the partitioning of the two meshes may not geometrically match. This is difficult to implement, but as @kronbichler and @peterrum mentioned, they have already provided this kind of functionality in the form of the `MGTwoLevelTransfer` class :-)

This patch makes use of this to implement the functions needed. The majority of the code in fact is due to the limitation that `MGTwoLevelTransfer` can only deal with `parallel::distributed::Vector` objects (see #16684). I decided that I was more interested in getting something to work for now than to make it super efficient, and so I just go with this limitation and copy data back and forth between the user-side vector types and the internally required `parallel::distributed::Vector` objects.

### CODE REVIEW

File: include/deal.II/numerics/vector_tools_interpolate.templates.h [hunk 1]
```diff
@@ -989,6 +1002,250 @@ namespace VectorTools
...
+    AssertDimension(dof_handler_coarse.n_dofs(), u_coarse.size());
+
+    using LAVector =
+      LinearAlgebra::distributed::Vector<typename VectorType::value_type>;
+
+    // Create and initialize a version of the coarse vector using the
+    // p::d::Vector type:
+    LAVector my_u_coarse;
+    InterpolateBetweenMeshes::create_vector(dof_handler_coarse, my_u_coarse);
+    my_u_coarse.update_ghost_values();
```
Comment (kronbichler) [inline include/deal.II/numerics/vector_tools_interpolate.templates.h:1131]: You don't need this, do you?
```suggestion
```

File: include/deal.II/numerics/vector_tools_interpolate.templates.h [hunk 2]
```diff
@@ -989,6 +1002,250 @@ namespace VectorTools
...
+    InterpolateBetweenMeshes::create_vector(dof_handler_coarse, my_u_coarse);
+    InterpolateBetweenMeshes::copy_locally_owned_data_from(u_coarse,
+                                                           my_u_coarse);
+    my_u_coarse.update_ghost_values();
+
+    // Then do the same for the fine vector, and initialize it with a copy
+    // of the source vector
+    LAVector my_u_fine;
+    InterpolateBetweenMeshes::create_vector(dof_handler_fine, my_u_fine);
+    my_u_fine.update_ghost_values();
```
Comment (kronbichler) [inline include/deal.II/numerics/vector_tools_interpolate.templates.h:1209]:
```suggestion
```

**Review by kronbichler — COMMENTED** [review]
I think this looks good, apart from some small comments.

I have an additional question: I think we can also address some of https://github.com/dealii/dealii/blob/54ef80696ba571cf90e31291fed9244b89d3688d/include/deal.II/fe/fe_tools.h#L621-L654 and similar functions in `FETools` with this infrastructure. This does not need to happen now, but I think it is good to have this in mind, to simplify life for users in this regard in the near future.

**Review by kronbichler — APPROVED** [review]

File: include/deal.II/numerics/vector_tools_interpolate.templates.h [hunk 3]
```diff
@@ -989,6 +1002,250 @@ namespace VectorTools
...
+    const DoFHandler<dim> &dof_handler_coarse,
+    const VectorType      &u_coarse,
+    const DoFHandler<dim> &dof_handler_fine,
+    const AffineConstraints<typename VectorType::value_type> &constraints_fine,
+    VectorType                                               &u_fine)
+  {
+    Assert(GridTools::have_same_coarse_mesh(dof_handler_fine,
+                                            dof_handler_coarse),
+           ExcMessage("The two DoF handlers must represent triangulations that "
+                      "have the same coarse meshes"));
```
Comment (peterrum) [inline include/deal.II/numerics/vector_tools_interpolate.templates.h:1190]: Let me point out that https://www.dealii.org/developer/doxygen/deal.II/classDataOutResample.html and https://www.dealii.org/developer/doxygen/deal.II/classMGTwoLevelTransferNonNested.html is for this purpose.

**Review by peterrum — APPROVED** [review]

File: include/deal.II/numerics/vector_tools_interpolate.templates.h (hunk 1 above)
Comment (bangerth) [reply include/deal.II/numerics/vector_tools_interpolate.templates.h:1131]: You're right, I don't!

**Comment by bangerth** [comment]
@kronbichler I removed the two lines. For the functions in `FETools`: Good grief, they look exactly like the functions in `VectorTools`. I had not remembered that we have this stuff twice. I'll leave a couple links in a separate PR (#16743).

**Review by kronbichler — APPROVED** [review]

**Pull request merged**